safekeeper/src/timeline.rs

```diff
@@ -614,23 +618,49 @@ impl Timeline {
             }
         }
 
-        // Create timeline directory.
-        fs::create_dir_all(&self.timeline_dir).await?;
+        // Create a temporary timeline directory
+        let (_tmp_dir, tli_dir_path) = create_temp_timeline_dir(conf, self.ttid).await?;
+
+        // Init the control file
+        let init_control_file = async {
+            let guard = shared_state.sk.state_mut();
+            let path = tli_dir_path.join(CONTROL_FILE_NAME);
+            let buf = guard.write_to_buf()?;
+            let mut control_file = File::create(&path)
+                .await
+                .with_context(|| format!("failed to create init control file at: {}", path))?;
+            control_file.write_all(&buf).await.with_context(|| {
+                format!(
+                    "failed to write safekeeper state into control file at: {}",
+                    path
+                )
+            })?;
+            control_file.flush().await.with_context(|| {
```

Comment:
flush is a rust thing which doesn't do fsync, it is no op in unix. That's why it is better to do such sensitive operations in a single place.

https://github.com/neondatabase/neon/issues/8927 has some more issues, so I started larger refactor of init and ended up with
https://github.com/neondatabase/neon/pull/9362
, please have a look.